File: src/main/java/de/cplaiz/activecraftcore/manager/PortalManager.java
package de.cplaiz.activecraftcore.manager;

import de.cplaiz.activecraftcore.ActiveCraftCore;
import de.cplaiz.activecraftcore.utils.config.Portal;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.block.EndGateway;

import java.util.Map;
import java.util.function.Predicate;

public class PortalManager {

    public static void create(String name, int x, int y, int z, World world, int to_x, int to_y, int to_z, World to_world) {
        ActiveCraftCore.getInstance().getPortalsConfig().set(name,
                Map.of("x", x, "y", y, "z", z, "world", world.getName(), "to_x", to_x, "to_y", to_y, "to_z", to_z, "to_world", to_world.getName())
                , true);
        Block block = world.getBlockAt(x, y, z);
        block.setType(Material.END_GATEWAY);
        EndGateway endGateway = (EndGateway) block.getState();
        endGateway.setExactTeleport(true);
        Location loc = new Location(to_world, to_x, to_y, to_z);
        endGateway.setExitLocation(loc);
        endGateway.setAge(-999999999);
        endGateway.update();
    }

    public static void destroy(String name) {
        Portal portal = ActiveCraftCore.getInstance().getPortalsConfig().getPortals().get(name);
        int portalx = portal.x();
        int portaly = portal.y();
        int portalz = portal.z();
        Block block = portal.world().getBlockAt(portalx, portaly, portalz);
        block.setType(Material.AIR);
        ActiveCraftCore.getInstance().getPortalsConfig().set(name, null, true);
    }

    public static void clean() {
        ActiveCraftCore.getInstance().getPortalsConfig().getPortals().values().stream()
                .filter(Predicate.not(portal -> portal.world() != null && portal.to_world() != null && portal.world().getBlockAt(portal.x(), portal.y(), portal.z()).getType() == Material.END_GATEWAY))
                .forEach(portal -> ActiveCraftCore.getInstance().getPortalsConfig().set(portal.name(), null, true));
    }
}

File: src/main/java/org/activecraft/activecraftcore/manager/PortalManager.kt
package org.activecraft.activecraftcore.manager

import org.activecraft.activecraftcore.ActiveCraftCore
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.block.EndGateway

object PortalManager {
    fun create(name: String, x: Int, y: Int, z: Int, world: World, toX: Int, toY: Int, toZ: Int, toWorld: World) {
        ActiveCraftCore.INSTANCE.portalsConfig.set(
            name,
            mapOf(
                "x" to x,
                "y" to y,
                "z" to z,
                "world" to world.name,
                "to_x" to toX,
                "to_y" to toY,
                "to_z" to toZ,
                "to_world" to toWorld.name
            ), true
        )
        val block = world.getBlockAt(x, y, z)
        block.type = Material.END_GATEWAY
        val endGateway = block.state as EndGateway
        endGateway.isExactTeleport = true
        val loc = Location(toWorld, toX.toDouble(), toY.toDouble(), toZ.toDouble())
        endGateway.exitLocation = loc
        endGateway.age = -999999999
        endGateway.update()
    }

    fun destroy(name: String) {
        val (_, portalX, portalY, portalZ, world) = ActiveCraftCore.INSTANCE.portalsConfig.portals[name]!!
        val block = world.getBlockAt(portalX, portalY, portalZ)
        block.type = Material.AIR
        ActiveCraftCore.INSTANCE.portalsConfig.set(name, null, true)
    }

    fun clean() {
        ActiveCraftCore.INSTANCE.portalsConfig.portals.values
            .filter {
                it.world.getBlockAt(it.x, it.y, it.z).type == Material.END_GATEWAY
            }
            .forEach { portal -> ActiveCraftCore.INSTANCE.portalsConfig.set(portal.name, null, true) }
    }
}